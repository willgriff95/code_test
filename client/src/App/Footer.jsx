import React from 'react'
import { withStyles } from '@material-ui/core/styles';
import TableCell from '@material-ui/core/TableCell';
import TableFooter from '@material-ui/core/TableFooter';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';

const styles = theme => ({

  })

const Footer = ({classes, rowsPerPage, page, data, handleChangePage, handleChangeRowsPerPage}) => {

    return(
    <TableFooter className={classes.pageNav}>
        <TableRow>
        <TableCell>
            <TablePagination
                labelRowsPerPage={''}
                rowsPerPageOptions={[]}
                component="div"
                count={data.length}
                rowsPerPage={rowsPerPage}
                page={page}
                backIconButtonProps={{
                    'aria-label': 'Previous Page',
                }}
                nextIconButtonProps={{
                    'aria-label': 'Next Page',
                }}
                onChangePage={handleChangePage}
                onChangeRowsPerPage={handleChangeRowsPerPage}
            />
        </TableCell>
        </TableRow>
    </TableFooter>
    )
}

export default withStyles(styles)(Footer)