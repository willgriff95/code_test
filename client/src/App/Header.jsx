import React from 'react'
import { Typography } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
    headerContainer: {
        background: '#0030ff',
    },
    headerText: {
      margin: '0 auto',
      width: '25%',
      color: 'white', 
      textAlign: 'left',
      letterSpacing: '15px',
      padding: '20px',
      lineHeight: '20px'
    }
  })

const Header = ({classes}) => (
    <div className={classes.headerContainer}>
        <Typography className={classes.headerText}>GENERAL  <br/>KNOWLEDGE_</Typography>
    </div>
);

export default withStyles(styles)(Header)