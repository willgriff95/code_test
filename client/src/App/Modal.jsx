import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Modal from '@material-ui/core/Modal';
import { Typography, Paper } from '@material-ui/core';
  
  function getModalStyle() {
    return {
      top: '40%',
      left: '35%',
      transform: `translate(-35%%, -35%)`,
    };
  }
  
  const styles = theme => ({
    paper: {
      textAlign: 'center',
      position: 'relative',
      width: theme.spacing.unit * 50,
      boxShadow: theme.shadows[5],
      padding: theme.spacing.unit * 4,
      outline: 'none',
      backgroundColor: '#0030ff',
      border: '10px solid white',
      color: 'white'
    },
    score: {
      fontSize: '50px',
      color: 'white',
      fontWeight: 'bold'
    },
    correctAndIncorrect: {
      fontSize: '12px',
      color: 'white'
    },
    yourScoreContainer: {
      borderRadius: 0,
      boxShadow: 'none',
      padding: '20px',
      color: '#0030ff',
      letterSpacing: '15px',
      position: 'absolute',
      width: '80%',
      top: '-20%'
    },
    yourScoreText: {
      fontWeight: 'bold',
      color: '#0030ff'
    }
  });

  
const SimpleModal = ({classes, state, handleClose, handleOpen}) => {
    return (
        <div>
            <Modal
            aria-labelledby="simple-modal-title"
            aria-describedby="simple-modal-description"
            open={state && state.modalOpen}
            onClose={handleClose}
            >
            <div style={getModalStyle()} className={classes.paper}>
            <Paper className={classes.yourScoreContainer}>
              <Typography className={classes.yourScoreText}>
                YOUR SCORE
              </Typography>
            </Paper>
              <Typography className={classes.score}>
                  {state && state.totalIncorrect_answers * 3}
              </Typography>
              <Typography className={classes.correctAndIncorrect}>
                  {state && state.totalIncorrect_answers} Correct
              </Typography>
              <Typography className={classes.correctAndIncorrect}>
                  {state && state.totalCorrect_answers} Incorrect
              </Typography>
              <SimpleModalWrapped />
            </div>
            </Modal>
        </div>
    )
}

SimpleModal.propTypes = {   
    classes: PropTypes.object.isRequired,
  };
  
  const SimpleModalWrapped = withStyles(styles)(SimpleModal);

  
export default SimpleModalWrapped;
