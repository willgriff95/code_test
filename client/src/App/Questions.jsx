import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import { Typography } from '@material-ui/core';

const styles = theme => ({
  root: {
    display: 'flex',
  },
  formControl: {
    display: 'block',
    width: '100%'
  },
  group: {
    margin: 0,
    position: 'relative',
    padding: '20px 20px 5px 20px',
    boxShadow: '10px 4px 29px -3px rgba(0,0,0,0.11);',
    border: '#e7e7e7 solid 1px'
  },
  questionContainer: {
    boxShadow: '10px 4px 29px -3px rgba(0,0,0,0.11);',
    width: '100%',
    border: '#e7e7e7 solid 1px',
    position: 'relative',
    top: '25px',
    left: '-30px',
    zIndex: '1'
  },
  questionText: {
    width: 'calc(100% - 60px)',   
    height: '60px',
    display: 'block',
    padding: '10px 10px 10px 20px',
    boxSizing: 'border-box',
    position: 'relative',
    float: 'left',
    backgroundColor: 'white',
    lineHeight: '20px',
    fontSize: '0.875rem',
    color: 'black'
  },
  questionNumber: {
    height: '60px',
    width: '60px',
    display: 'block',
    float: 'left',
    padding: '20px',
    boxSizing: 'border-box',
    position: 'relative',
    backgroundColor: '#0030ff',
    color: 'white'
  }
});

class Questions extends React.Component {
  render() {
    const { classes, questionData, radioButtonChange, index, questionIndex } = this.props;
    const questionNew = <div dangerouslySetInnerHTML={{__html: `<Typography>${questionData && questionData.question}</Typography>`}} />

    return (
      <div className={classes.root}>
        <FormControl component="fieldset" className={classes.formControl}>
        <FormLabel component="legend" className={classes.questionContainer}>
            <Typography className={classes.questionNumber}>
            {(questionIndex + 1).toString().length > 1 ? 
              `${questionIndex + 1}_`
              :
              `0${questionIndex + 1}_`
          }
            </Typography>
            <div className={classes.questionText}>
                {questionNew}
            </div>
        </FormLabel>
          <RadioGroup
            aria-label={index}
            className={classes.group}
            questionnumber={index}
          >
          {questionData && questionData.answers && questionData.answers.map((answer, index) => {
            const answerNew = <div dangerouslySetInnerHTML={{__html: `<div>${answer}</div>`}} />
            return (
              <FormControlLabel 
                key={index}  
                questionnumber={index} 
                value={answer} 
                onChange={radioButtonChange} 
                control={<Radio color="primary" />} 
                label={answerNew}
              />
            )
          })}
          </RadioGroup>
        </FormControl>
      </div>
    );
  }
}

Questions.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Questions);