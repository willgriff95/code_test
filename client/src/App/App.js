import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';

import TableRow from '@material-ui/core/TableRow';
import Questions from './Questions';
import SimpleModalWrapped from './Modal';
import Header from './Header'
import Footer from './Footer'
const styles = theme => ({
  root: {
    width: '100%',
    marginTop: theme.spacing.unit * 3,
  },
  table: {
    width: '550px',
    margin: '0 auto'
  },
  tableWrapper: {
    overflowX: 'auto',
  },
  questions: {
    height: '100%',
    display: 'block',
    margin: '0 auto 5px auto',
    backgroundColor: 'white'
  },
  tableBody: {
    width: '50%'
  }, 
  questionsCell: {
    borderBottom: 0,
    width: '100%',
    display: 'block',
    position: 'relative',
    padding: '0 !important'
  },
  pageNav: {
  }
});

class App extends React.Component {
  state = {
    rows: [],
    page: 0,
    data: [
    ],
    selected: [],
    rowsPerPage: 3,
    modalOpen: false,
    totalIncorrect_answers: 0,
    totalCorrect_answers: 0
  }

  componentDidMount() { 
    this.getQuestions()
  }

  getQuestions = () => {

    // fetching from an api
    // fetch('https://opentdb.com/api.php?amount=20&difficulty=hard&type=multiple')

    // OR
    
    // fetching locally
    fetch('/api/questions')
    .then(res => res.json())
    .then(data => this.setState({ data: data.results }, () => {
        this.getAnswers()
    }))

  }

  handleChangePage = (event, page) => {
    this.setState({ page })
  }

  handleChangeRowsPerPage = event => {
    this.setState({ page: 0, rowsPerPage: event.target.value })
  }

  getAnswers = () => {
    
    this.state.data.forEach(((questionData, index) => {
      let answers = []
      questionData.incorrect_answers && questionData.incorrect_answers.forEach((incorrect_answer, index) => {
        answers.push(incorrect_answer)
      })
      answers.push(questionData.correct_answer)
      answers = this.shuffle(answers)
      const tempData = this.state.data  
      tempData[index].answers = answers
      this.setState({
        data: tempData
      })
    }))
  }

  shuffle = (answers) => {
    var currentIndex = answers.length, temporaryValue, randomIndex;
  
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = answers[currentIndex];
      answers[currentIndex] = answers[randomIndex];
      answers[randomIndex] = temporaryValue;
    }
  
    return answers;
  }

  radioButtonChange = (questionIndex, event) => {
    const tempData = this.state.data  
    tempData[questionIndex].selectedAnswer = event.target.value
    this.setState({ value: event.target.value })
    this.calculateTotalAnswers(event)
    this.checkAllquestionsAnswered()
  }

  checkAllquestionsAnswered = () => {
    let numberOfQuestions = this.state.data.length
    let allQuestionsAnswered = 0
    this.state.data.forEach((questionData) => {
      if(questionData.selectedAnswer) {
        allQuestionsAnswered ++ 
      }
    })
    if (numberOfQuestions === allQuestionsAnswered) {
      this.handleOpen()
    }
  }

  calculateTotalAnswers = (event) => {
    let totalCorrect_answers = 0
    let totalIncorrect_answers = 0
    this.state.data.forEach(((questionData, index) => {
      if(questionData.selectedAnswer === questionData.correct_answer) {
        const tempData = this.state.data  
        tempData[index].answerCorrect = true
        tempData[index].answerIncorrect = false
        this.setState({
          data: tempData
        })
      } else {
        const tempData = this.state.data  
        tempData[index].answerIncorrect = true
        tempData[index].answerCorrect = false
        this.setState({
          data: tempData
        })
      }
      if (questionData.answerCorrect){
        totalCorrect_answers ++
      } else {
        totalIncorrect_answers ++
      }
    }))

    this.setState({
      totalCorrect_answers,
      totalIncorrect_answers
    })
  }

  handleOpen = () => {
    this.setState({ modalOpen: true })
  }

  handleClose = () => {
    this.setState({ modalOpen: false })
  }


  render() {
    const { classes } = this.props;
    const { rowsPerPage, page, data } = this.state;
    return (
      <div>
        <Header />
        <div className={classes.tableWrapper}>
          <Table className={classes.table}>
            <TableBody  className={classes.tableBody}>
            {data 
            .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
            .map((questionData, index) => {
              const questionIndex = page * rowsPerPage + index
                return (
                  <TableRow 
                    key={index} 
                    className={classes.questions}
                  >
                    <TableCell className={classes.questionsCell}>
                      <Questions 
                        questionData={questionData} 
                        state={this.state} 
                        value={this.state.value} 
                        radioButtonChange={(e) => this.radioButtonChange(questionIndex, e)} 
                        questionIndex={questionIndex} 
                      />
                    </TableCell>
                  </TableRow>
                )
              })}
            </TableBody>
            <Footer 
              data={data} 
              rowsPerPage={rowsPerPage} 
              page={page} 
              handleChangePage={this.handleChangePage}  
              handleChangeRowsPerPage={this.handleChangeRowsPerPage}
            />
          </Table>
          <SimpleModalWrapped 
            state={this.state} 
            handleOpen={this.handleOpen} 
            handleClose={this.handleClose}
          />
        </div>
      </div>
    );
  }
}

App.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(App);
