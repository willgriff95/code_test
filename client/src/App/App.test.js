import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import axios from "axios";
import { create } from "react-test-renderer";
import * as renderer from 'react-test-renderer';
import questionsJson from '../../../Questions/Questions.json'
it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
})

it('render correctly App component', () => {  
  const AppComponent = renderer.create(<App />).toJSON();
  expect(AppComponent).toMatchSnapshot();
});

jest.mock("axios");
describe("App component", () => {
  it("shows a list of Questions", async () => {
    const response = questionsJson
    axios.get.mockResolvedValue(response);
    const component = create(<App />);
    const instance = component.getInstance();
    await instance.componentDidMount();
  });
});