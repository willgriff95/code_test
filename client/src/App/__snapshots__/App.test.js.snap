// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly App component 1`] = `
<div>
  <div
    className="Header-headerContainer-112"
  >
    <p
      className="MuiTypography-root-114 MuiTypography-body1-123 Header-headerText-113"
    >
      GENERAL  
      <br />
      KNOWLEDGE_
    </p>
  </div>
  <div
    className="App-tableWrapper-107"
  >
    <table
      className="MuiTable-root-150 App-table-106"
    >
      <tbody
        className="MuiTableBody-root-151 App-tableBody-109"
      />
      <tfoot
        className="MuiTableFooter-root-152"
      >
        <tr
          className="MuiTableRow-root-153 MuiTableRow-footer-157"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root-158 MuiTableCell-footer-161"
          >
            <div
              className="MuiTablePagination-root-170"
            >
              <div
                className="MuiToolbar-root-180 MuiToolbar-regular-182 MuiToolbar-gutters-181 MuiTablePagination-toolbar-171"
              >
                <div
                  className="MuiTablePagination-spacer-172"
                />
                <span
                  className="MuiTypography-root-114 MuiTypography-caption-124 MuiTypography-colorInherit-143 MuiTablePagination-caption-173"
                >
                  0-0 of 0
                </span>
                <div
                  className="MuiTablePagination-actions-179"
                >
                  <button
                    aria-label="Previous Page"
                    className="MuiButtonBase-root-190 MuiButtonBase-disabled-191 MuiIconButton-root-184 MuiIconButton-colorInherit-185 MuiIconButton-disabled-188"
                    disabled={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="-1"
                    type="button"
                  >
                    <span
                      className="MuiIconButton-label-189"
                    >
                      <svg
                        aria-hidden="true"
                        className="MuiSvgIcon-root-193"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                        />
                      </svg>
                    </span>
                  </button>
                  <button
                    aria-label="Next Page"
                    className="MuiButtonBase-root-190 MuiButtonBase-disabled-191 MuiIconButton-root-184 MuiIconButton-colorInherit-185 MuiIconButton-disabled-188"
                    disabled={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="-1"
                    type="button"
                  >
                    <span
                      className="MuiIconButton-label-189"
                    >
                      <svg
                        aria-hidden="true"
                        className="MuiSvgIcon-root-193"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
    <div />
  </div>
</div>
`;
